pkg/storagev2/metrics.go

```diff
@@ -0,0 +1,320 @@
+		m.metrics.IndexStoreCalls.WithLabelValues("Has", "success").Inc()
+	} else {
+		m.metrics.IndexStoreCalls.WithLabelValues("Has", "failure").Inc()
+	}
+	return has, err
+}
+
+// GetSize implements the TxStore interface.
+func (m txIndexStoreWithMetrics) GetSize(key Key) (int, error) {
+	now := time.Now()
+	size, err := m.TxStore.GetSize(key)
+	m.metrics.IndexStoreCallsDuration.WithLabelValues("GetSize").Observe(float64(time.Since(now).Nanoseconds()))
+	if err == nil || errors.Is(err, ErrNotFound) {
+		m.metrics.IndexStoreCalls.WithLabelValues("GetSize", "success").Inc()
+	} else {
+		m.metrics.IndexStoreCalls.WithLabelValues("GetSize", "failure").Inc()
+	}
+	return size, err
+}
+
+// Iterate implements the TxStore interface.
+func (m txIndexStoreWithMetrics) Iterate(query Query, fn IterateFn) error {
+	now := time.Now()
+	err := m.TxStore.Iterate(query, fn)
+	m.metrics.IndexStoreCallsDuration.WithLabelValues("Iterate").Observe(float64(time.Since(now).Nanoseconds()))
+	if err == nil || errors.Is(err, ErrNotFound) {
+		m.metrics.IndexStoreCalls.WithLabelValues("Iterate", "success").Inc()
+	} else {
+		m.metrics.IndexStoreCalls.WithLabelValues("Iterate", "failure").Inc()
+	}
+	return err
+}
+
+// Count implements the TxStore interface.
+func (m txIndexStoreWithMetrics) Count(key Key) (int, error) {
+	now := time.Now()
+	cnt, err := m.TxStore.Count(key)
+	m.metrics.IndexStoreCallsDuration.WithLabelValues("Count").Observe(float64(time.Since(now).Nanoseconds()))
+	if err == nil || errors.Is(err, ErrNotFound) {
+		m.metrics.IndexStoreCalls.WithLabelValues("Count", "success").Inc()
+	} else {
+		m.metrics.IndexStoreCalls.WithLabelValues("Count", "failure").Inc()
+	}
+	return cnt, err
+}
+
+// Put implements the TxStore interface.
+func (m txIndexStoreWithMetrics) Put(item Item) error {
+	now := time.Now()
+	err := m.TxStore.Put(item)
+	m.metrics.IndexStoreCallsDuration.WithLabelValues("Put").Observe(float64(time.Since(now).Nanoseconds()))
+	if err == nil || errors.Is(err, ErrNotFound) {
+		m.metrics.IndexStoreCalls.WithLabelValues("Put", "success").Inc()
+	} else {
+		m.metrics.IndexStoreCalls.WithLabelValues("Put", "failure").Inc()
+	}
+	return err
+}
+
+// Delete implements the TxStore interface.
+func (m txIndexStoreWithMetrics) Delete(item Item) error {
+	now := time.Now()
+	err := m.TxStore.Delete(item)
+	m.metrics.IndexStoreCallsDuration.WithLabelValues("Delete").Observe(float64(time.Since(now).Nanoseconds()))
+	if err == nil || errors.Is(err, ErrNotFound) {
+		m.metrics.IndexStoreCalls.WithLabelValues("Delete", "success").Inc()
+	} else {
+		m.metrics.IndexStoreCalls.WithLabelValues("Delete", "failure").Inc()
+	}
+	return err
+}
+
+var _ TxChunkStore = (*txChunkStoreWithMetrics)(nil)
+
+// txChunkStoreWithMetrics wraps TxChunkStore and adds metrics.
+type txChunkStoreWithMetrics struct {
+	TxChunkStore
+
+	metrics metrics
+}
+
+// Commit implements the TxChunkStore interface.
+func (m txChunkStoreWithMetrics) Commit() error {
+	now := time.Now()
```

Comment:
We can also have a metric that will calculate the total time since the inception of this transaction. We can have a New method that creates this initial timestamp and then check it in Commit and Rollback to report the TTL of the transaction as a whole. This could be useful information.

Reply:
Done.